### PR TITLE
Restrict input argument as orderable in array_sort, array_sort_desc

### PR DESCRIPTION
**Pull request by duanmeng**
In Presto, input to `array_sort`, and `array_sort_desc` are restricted to
orderable types. For example, MAP type is not orderable.

This PR uses the `orderableTypeVariable` to apply the restriction
to the input argument of array_sort. This restriction applies to the input
argument type of normal array_sort, and the return type of the lambda
of array_sort with a transform lambda, not the array_sort function
with a custom comparator.

For more details please check the blog https://velox-lib.io/blog/array-sort/

```SQL
presto> SELECT array_sort(ARRAY [map(array['a', 'b', 'c'], array[1, 2, 3])]);]
...
Expected: array_sort(array(E)) E:orderable, ...

presto> SELECT array_sort(ARRAY [map(array['a', 'b', 'c'], array[1, 2, 3])], 
(x, y) -> IF(cardinality(x) < cardinality(y), -1, IF(cardinality(x) = cardinality(y), 0, 1)));
       _col0
-------------------
 [{a=1, b=2, c=3}]
(1 row)
```

Part of #6718

### CODE REVIEW

File: velox/functions/prestosql/ArraySort.cpp [hunk 1]
```diff
@@ -374,13 +374,13 @@ std::vector<std::shared_ptr<exec::FunctionSignature>> signatures(
...
       exec::FunctionSignatureBuilder()
-          .typeVariable("T")
+          .orderableTypeVariable("T")
           .returnType("array(T)")
           .argumentType("array(T)")
           .build(),
       // array(T), function(T,U), boolean -> array(T)
       exec::FunctionSignatureBuilder()
-          .typeVariable("T")
+          .orderableTypeVariable("T")
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArraySort.cpp:383]: Here, T doesn't have to be orderable, but U has to be.

For example, it should be possible to sort array of maps by cardinality of these maps.

File: velox/functions/prestosql/tests/ArraySortTest.cpp [hunk 2]
```diff
@@ -229,7 +229,8 @@ class ArraySortTest : public FunctionBaseTest,
         test<StringView>();
         break;
       case TypeKind::MAP:
-        test<TestMapType>();
+        VELOX_ASSERT_THROW(
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArraySortTest.cpp:232]: Let's remove all test code related to maps and just have a dedicate test to verify that sorting arrays of maps is not allowed.

**Review by mbasmanova — COMMENTED** [review]
Thanks. Looks good % some comments.

File: velox/functions/prestosql/ArraySort.cpp (hunk 1 above)
Comment (duanmeng) [reply velox/functions/prestosql/ArraySort.cpp:383]: For array_sort with a transform lambda, the Presto still checks the T, I describe it in the PR description,
```SQL
presto> SELECT array_sort(ARRAY [map(array['a', 'b', 'c'], array[1, 2, 3])], 
(x, y) -> cardinality(x) < cardinality(y));
...
Expected: array_sort(array(E)) E:orderable, ...
```

Comment (mbasmanova) [reply velox/functions/prestosql/ArraySort.cpp:383]: This sounds like a bug. Let's create an issue in prestodb repo.

File: velox/functions/prestosql/tests/ArraySortTest.cpp [hunk 3]
```diff
@@ -229,7 +229,7 @@ class ArraySortTest : public FunctionBaseTest,
         test<StringView>();
         break;
       case TypeKind::MAP:
-        test<TestMapType>();
+        // Ignore test for MAP for it is not an orderable type.
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArraySortTest.cpp:232]: TestMapType and all the code using it is now dead code. Let's remove it.

File: velox/functions/prestosql/tests/ArraySortTest.cpp [hunk 4]
```diff
@@ -583,6 +583,22 @@ TEST_F(ArraySortTest, lambda) {
...
+      "Scalar function signature is not supported"_sv;
+  auto testFail = [&](const std::string& name, const std::string& lambdaExpr) {
+    VELOX_ASSERT_THROW(
+        evaluate(fmt::format("{}(c0, {})", name, lambdaExpr), data),
+        kErrorMessage);
+    VELOX_ASSERT_THROW(
+        evaluate(fmt::format("{}(c0)", name), data), kErrorMessage);
+  };
+
+  testFail("array_sort", "x -> x");
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArraySortTest.cpp:598]: Can we also test non-lambda versions of the functions?

File: velox/functions/prestosql/tests/ArraySortTest.cpp [hunk 5]
```diff
@@ -583,6 +583,22 @@ TEST_F(ArraySortTest, lambda) {
       "(x, y) -> if(length(x) < length(y), 1, if(length(x) = length(y), 0, -1))");
 }
 
+TEST_F(ArraySortTest, failOnMapTypeSort) {
+  auto data = makeRowVector({makeDataArray<TestMapType>({2, 1, 0})});
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArraySortTest.cpp:587]: we can just use makeAllNullMapVector(10, INTEGER(), VARCHAR())

File: velox/functions/prestosql/tests/ArraySortTest.cpp (hunk 4 above)
Comment (duanmeng) [reply velox/functions/prestosql/tests/ArraySortTest.cpp:598]: The function testFail test both non-lambda and lambda version.
```C++
VELOX_ASSERT_THROW(
        evaluate(fmt::format("{}(c0)", name), data), kErrorMessage);
```

File: velox/functions/prestosql/tests/ArraySortTest.cpp (hunk 5 above)
Comment (duanmeng) [reply velox/functions/prestosql/tests/ArraySortTest.cpp:587]: Got it, thanks.

**Review by mbasmanova — COMMENTED** [review]
Thank you for iterating. A couple remaining comments.

File: velox/functions/prestosql/ArraySort.cpp [hunk 6]
```diff
@@ -374,14 +374,14 @@ std::vector<std::shared_ptr<exec::FunctionSignature>> signatures(
   std::vector<std::shared_ptr<exec::FunctionSignature>> signatures = {
       // array(T) -> array(T)
       exec::FunctionSignatureBuilder()
-          .typeVariable("T")
+          .orderableTypeVariable("T")
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArraySort.cpp:377]: Consider updating the documentation to add this restriction.

File: velox/functions/prestosql/tests/ArraySortTest.cpp (hunk 4 above)
Comment (mbasmanova) [reply velox/functions/prestosql/tests/ArraySortTest.cpp:598]: Got it. I missed this. Thank you for clarifying.

File: velox/functions/prestosql/tests/ArraySortTest.cpp (hunk 3 above)
Comment (duanmeng) [reply velox/functions/prestosql/tests/ArraySortTest.cpp]: Done

File: velox/functions/prestosql/ArraySort.cpp (hunk 6 above)
Comment (duanmeng) [reply velox/functions/prestosql/ArraySort.cpp:377]: I've updated all the places related to array_sort in the document.

File: velox/functions/prestosql/ArraySort.cpp (hunk 1 above)
Comment (duanmeng) [reply velox/functions/prestosql/ArraySort.cpp]: Got it, I've updated the PR description too.

**Review by mbasmanova — APPROVED** [review]
Thank you.

File: velox/functions/prestosql/tests/ArraySortTest.cpp [hunk 7]
```diff
@@ -583,6 +542,21 @@ TEST_F(ArraySortTest, lambda) {
...
+  auto testFail = [&](const std::string& name, const std::string& lambdaExpr) {
+    VELOX_ASSERT_THROW(
+        evaluate(fmt::format("{}(c0, {})", name, lambdaExpr), data),
+        "(1 vs. 0)");
+    VELOX_ASSERT_THROW(
+        evaluate(fmt::format("{}(c0)", name), data),
+        "Scalar function signature is not supported");
+  };
+
+  testFail("array_sort", "x -> x");
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArraySortTest.cpp:556]: nit: It is confusing that this call passes a lambda. This makes it look like we are only testing lambda version of the function. Since the lambda is the same in all calls, consider removing lambda argument and hard-coding it inside testFail.

File: velox/functions/prestosql/tests/ArraySortTest.cpp [hunk 8]
```diff
@@ -583,6 +542,21 @@ TEST_F(ArraySortTest, lambda) {
       "(x, y) -> if(length(x) < length(y), 1, if(length(x) = length(y), 0, -1))");
 }
 
+TEST_F(ArraySortTest, failOnMapTypeSort) {
+  auto data = makeRowVector({makeAllNullMapVector(10, INTEGER(), VARCHAR())});
+  auto testFail = [&](const std::string& name, const std::string& lambdaExpr) {
+    VELOX_ASSERT_THROW(
+        evaluate(fmt::format("{}(c0, {})", name, lambdaExpr), data),
+        "(1 vs. 0)");
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArraySortTest.cpp:550]: Let's verify more of the error message. (1 vs. 0) is not very descriptive.

Comment (mbasmanova) [reply velox/functions/prestosql/tests/ArraySortTest.cpp:550]: I wonder whether we are testing the test framework here. Aren't there errors thrown while parsing the expression, not when trying to evaluate a function. If so, I wonder if it is better to just remove these tests.

Comment (duanmeng) [reply velox/functions/prestosql/tests/ArraySortTest.cpp:550]: Yes, this test case is wrong, I am fixing it.

Comment (duanmeng) [reply velox/functions/prestosql/tests/ArraySortTest.cpp:550]: I should use `makeAllNullMapVector` to create an ArrayVector before putting it into the RowVector.

File: velox/functions/prestosql/tests/ArraySortTest.cpp (hunk 7 above)
Comment (duanmeng) [reply velox/functions/prestosql/tests/ArraySortTest.cpp]: Fixed.

File: velox/functions/prestosql/tests/ArraySortTest.cpp [hunk 9]
```diff
@@ -583,6 +542,28 @@ TEST_F(ArraySortTest, lambda) {
...
 
+TEST_F(ArraySortTest, failOnMapTypeSort) {
+  static const StringView kErrorMessage =
+      "Scalar function signature is not supported"_sv;
+  std::vector<std::pair<std::string, std::optional<int64_t>>> a{
+      {"a", {1}}, {"b", {2}}};
+  std::vector<std::pair<std::string, std::optional<int64_t>>> b{
+      {"c", {3}}, {"d", {4}}};
+  auto arrayOfMapVector =
+      makeArrayOfMapVector<std::string, int64_t>({{a, b}, {std::nullopt}});
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArraySortTest.cpp:553]: Why makeAllNullMapVector doesn't work?

File: velox/functions/prestosql/tests/ArraySortTest.cpp [hunk 10]
```diff
@@ -583,6 +542,28 @@ TEST_F(ArraySortTest, lambda) {
...
+  auto data = makeRowVector({arrayOfMapVector});
+  auto testFail = [&](const std::string& name) {
+    VELOX_ASSERT_THROW(
+        evaluate(fmt::format("{}(c0, {})", name, "x -> x"), data),
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArraySortTest.cpp:557]: "{}(c0, x -> x)"

File: velox/functions/prestosql/tests/ArraySortTest.cpp [hunk 11]
```diff
@@ -583,6 +542,24 @@ TEST_F(ArraySortTest, lambda) {
       "(x, y) -> if(length(x) < length(y), 1, if(length(x) = length(y), 0, -1))");
 }
 
+TEST_F(ArraySortTest, failOnMapTypeSort) {
+  auto defaultPool = memory::addDefaultLeafMemoryPool();
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArraySortTest.cpp:546]: this shouldn't be needed; use pool()

File: velox/functions/prestosql/tests/ArraySortTest.cpp (hunk 10 above)
Comment (duanmeng) [reply velox/functions/prestosql/tests/ArraySortTest.cpp]: Fixed

File: velox/functions/prestosql/tests/ArraySortTest.cpp [hunk 12]
```diff
@@ -583,6 +542,23 @@ TEST_F(ArraySortTest, lambda) {
       "(x, y) -> if(length(x) < length(y), 1, if(length(x) = length(y), 0, -1))");
 }
 
+TEST_F(ArraySortTest, failOnMapTypeSort) {
+  static const StringView kErrorMessage =
+      "Scalar function signature is not supported"_sv;
+  auto arrayOfMapVector = BaseVector::createNullConstant(
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArraySortTest.cpp:548]: this variable is not needed

File: velox/functions/prestosql/tests/ArraySortTest.cpp [hunk 13]
```diff
@@ -583,6 +542,23 @@ TEST_F(ArraySortTest, lambda) {
       "(x, y) -> if(length(x) < length(y), 1, if(length(x) = length(y), 0, -1))");
 }
 
+TEST_F(ArraySortTest, failOnMapTypeSort) {
+  static const StringView kErrorMessage =
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArraySortTest.cpp:546]: Why StringView? std::string

File: velox/functions/prestosql/tests/ArraySortTest.cpp (hunk 9 above)
Comment (duanmeng) [reply velox/functions/prestosql/tests/ArraySortTest.cpp]: It works and is convenient.

File: velox/functions/prestosql/tests/ArraySortTest.cpp (hunk 13 above)
Comment (duanmeng) [reply velox/functions/prestosql/tests/ArraySortTest.cpp:546]: I thought people preferred `StringView` to `std::string` for better performance.

File: velox/functions/prestosql/tests/ArraySortTest.cpp (hunk 11 above)
Comment (duanmeng) [reply velox/functions/prestosql/tests/ArraySortTest.cpp]: Yes, I found it and force pushed.

File: velox/functions/prestosql/tests/ArraySortTest.cpp (hunk 13 above)
Comment (duanmeng) [reply velox/functions/prestosql/tests/ArraySortTest.cpp]: I see, the `evaluate` method's param is `std::string`. Fixed.

Comment (mbasmanova) [reply velox/functions/prestosql/tests/ArraySortTest.cpp]: StringView should be used only for data stored in Velox vectors.

std::string_view can be used otherwise. One thing to keep in mind is that neither StringView nor std::string_view own the backing string. Hence, one needs to make sure that the backing string stays alive for as long as the view is alive.

Comment (duanmeng) [reply velox/functions/prestosql/tests/ArraySortTest.cpp]: Got it, thanks a lot.

**Comment by mbasmanova** [comment]
CI is red. Looks like we need to update AggregationFuzzer to respect "orderable" constraint.

```
I20231006 10:10:20.467994 202359 AggregationFuzzer.cpp:709] ==============================> Started iteration 18 (seed: 4059922645)
terminate called after throwing an instance of 'facebook::velox::VeloxUserError'
  what():  Exception: VeloxUserError
Error Source: USER
Error Code: INVALID_ARGUMENT
Reason: Scalar function signature is not supported: array_sort(ARRAY<MAP<INTERVAL DAY TO SECOND,DOUBLE>>). Supported signatures: (array(T)) -> array(T), (array(T),constant function(T,U)) -> array(T), (array(T),constant function(T,T,bigint)) -> array(T).
Retriable: False
Function: resolveScalarFunctionType
File: ../../velox/parse/TypeResolver.cpp
Line: 99
```

**Comment by duanmeng** [comment]
> CI is red. Looks like we need to update AggregationFuzzer to respect "orderable" constraint.

@mbasmanova We could add this constraint in `ArgumentTypeFuzzer::determineUnboundedTypeVariables`, in which a random type is generated.

File: velox/expression/tests/ArgumentTypeFuzzer.cpp [hunk 14]
```diff
@@ -48,7 +48,24 @@ void ArgumentTypeFuzzer::determineUnboundedTypeVariables() {
     // Random randomType() never generates unknown here.
     // TODO: we should extend randomType types and exclude unknown based
     // on variableInfo.
-    bindings_[variableName] = randType();
+    // Respect orderable/comparable constraint.
+    TypePtr typePtr = nullptr;
```
Comment (mbasmanova) [inline velox/expression/tests/ArgumentTypeFuzzer.cpp:52]: Let's move this to a separate PR and add a test. See velox/expression/tests/ArgumentTypeFuzzerTest.cpp

File: velox/expression/tests/ArgumentTypeFuzzer.cpp [hunk 15]
```diff
@@ -48,7 +48,24 @@ void ArgumentTypeFuzzer::determineUnboundedTypeVariables() {
...
+    while (true) {
```
Comment (mbasmanova) [inline velox/expression/tests/ArgumentTypeFuzzer.cpp:53]: This may take a very long time... We need to put a limit on the number of iterations or find a better way altogether. 

Perhaps, create a custom version of randType that skips MAPs.

```
TypePtr randType(
    FuzzerGenerator& rng,
    const std::vector<TypePtr>& scalarTypes,
    int maxDepth) {
  const int numScalarTypes = scalarTypes.size();
  // Should we generate a scalar type?
  if (maxDepth <= 1 || rand<bool>(rng)) {
    return scalarTypes[rand<uint32_t>(rng) % numScalarTypes];
  }
  switch (rand<uint32_t>(rng) % 2) {
    case 0:
      return ARRAY(randType(rng, scalarTypes, maxDepth - 1));
    default:
      return randRowType(rng, scalarTypes, maxDepth - 1);
  }
}
```

Comment (duanmeng) [reply velox/expression/tests/ArgumentTypeFuzzer.cpp:53]: Got it, thanks. I will revert this commit and file another PR adding a test according to your guidance.

Comment (mbasmanova) [reply velox/expression/tests/ArgumentTypeFuzzer.cpp]: @duanmeng I'll be out till next Wed. 

Our custom function may look like this.

```
TypePtr randOrderableType(
    FuzzerGenerator& rng,
    int maxDepth = 5) {
  // Should we generate a scalar type?
  if (maxDepth <= 1 || rand<bool>(rng)) {
    return randType(rng, 0);
  }

  // ARRAY or ROW?
  if (rand<bool>(rng)) {
      return ARRAY(randOrderableType(rng, maxDepth - 1));
  } 

  int numFields = 1 + rand<uint32_t>(rng) % 7;
  std::vector<std::string> names;
  std::vector<TypePtr> fields;
  for (int i = 0; i < numFields; ++i) {
    names.push_back(fmt::format("f{}", i));
    fields.push_back(randOrderableType(rng, maxDepth - 1));
  }
  return ROW(std::move(names), std::move(fields));  
}
```

**Comment by duanmeng** [comment]
@mbasmanova Hi Masha, I've rebased and added a mitigation purpose function `array_sort_relax`. I ran the aggregation fuzzer for 1800s using param `--seed 0 --duration_sec 1800 --logtostderr=1 --minloglevel=0 --repro_persist_path=./tmp/aggregate_fuzzer_repro`, it succeed.

**Comment by mbasmanova** [comment]
> @mbasmanova Hi Masha, I've rebased and added a mitigation purpose function `array_sort_relax`. I ran the aggregation fuzzer for 1800s using param `--seed 0 --duration_sec 1800 --logtostderr=1 --minloglevel=0 --repro_persist_path=./tmp/aggregate_fuzzer_repro`, it succeed.

@duanmeng This is nice. Let's update PR description to clarify that this change applies to array_sort_desc as well.

Let's also update rewriteArraySortCall to check the type of the transform.

Maybe rename array_sort_relax to $internal$array_sort to make it clear this is for internal usage only. I expect this function to fail if inputs contain nested nulls. Wondering why this didn't happen in the fuzzer run. What if you run the fuzzer with --only array_agg? I think we need to make sure that $internal$array_sort can handle all types and values including complex type values with nested nulls. What do you think?

**Comment by mbasmanova** [comment]
> $internal$array_sort 

An alternative solution might be to apply_sort after converting input to JSON: array_sort(json_format(cast(x as json)).

**Comment by duanmeng** [comment]
> An alternative solution might be to apply_sort after converting input to JSON: array_sort(json_format(cast(x as json)).

@mbasmanova This is more appropriate and cleaner (I should think more before coding :) ). 
I just checked all the result types of the functions in `customVerificationFunctions` that use `array_sort` to transform results, which are array or map and could be cast to `json`.(`array_agg, set_agg, set_union, map_agg, map_union, map_union_sum,multimap_agg`).  I will use this way to update this PR and rerun the fuzzer test.

PS: `array_agg_partial`, `array_agg_merge`, and `array_agg_merge_extract` seem useless, should we remove them?

**Comment by mbasmanova** [comment]
> PS: array_agg_partial, array_agg_merge, and array_agg_merge_extract seem useless, should we remove them?

There are so-called companion functions. See https://github.com/facebookincubator/velox/pull/4566

These are not currently used though.

I do see lots of issues when testing companion functions in the fuzzer. I was thinking to modify the registration APIs to allow to exclude all companion functions. CC: @kagamiori

**Comment by duanmeng** [comment]
> An alternative solution might be to apply_sort after converting input to JSON: array_sort(json_format(cast(x as json)).

@mbasmanova Hi Masha, it seems we could not use JSON type as DuckDB does not support it. I am trying other transform functions, if not work would fall back to the specific functions way.
```C++
terminating with uncaught exception of type std::runtime_error: unsupported type for duckdb -> velox conversion: JSON
```

**Comment by duanmeng** [comment]
@mbasmanova Hi Masha, I've updated this PR with the following changes, could you please help to take a look? Thanks.
- Use `internal$array_sort` as the specific internal function name.(`$internal$array_sort` is not supported by the `duckdb-libpq_query`)
- Add the return type of the transform (comparison) check in the `rewriteArraySortCall`.
- Update the PR description and title to clarify that this change applies to array_sort_desc as well.

**Comment by mbasmanova** [comment]
@duanmeng 

> $internal$array_sort is not supported by the duckdb-libpq_query

You may need to put $internal$array_sort in double quotes: `"$internal$array_sort"(x)`

**Comment by mbasmanova** [comment]
> $internal$array_sort

I wonder if maybe a better name for this function would be $internal$canonicalize, because this function transforms an array into canonical form that allows for comparison.

File: velox/functions/prestosql/ArraySort.cpp [hunk 16]
```diff
@@ -464,4 +489,12 @@ VELOX_DECLARE_STATEFUL_VECTOR_FUNCTION(
     signatures(false),
     createDesc);
 
+// Add an internal version of array_sort for used by AggregationFuzzerTest to
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArraySort.cpp:492]: typos:

```
An internal function to canonicalize an array to allow for comparisons. Used in AggregationFuzzerTest.
```

File: velox/functions/prestosql/registration/ArrayFunctionsRegistration.cpp [hunk 17]
```diff
@@ -122,6 +122,8 @@ void registerArrayFunctions(const std::string& prefix) {
   VELOX_REGISTER_VECTOR_FUNCTION(udf_array_sort, prefix + "array_sort");
   VELOX_REGISTER_VECTOR_FUNCTION(
       udf_array_sort_desc, prefix + "array_sort_desc");
+  VELOX_REGISTER_VECTOR_FUNCTION(
```
Comment (mbasmanova) [inline velox/functions/prestosql/registration/ArrayFunctionsRegistration.cpp:125]: Perhaps, move this registration into AggregationFuzzerTest. This function should not be available in production setup.

File: velox/functions/prestosql/ArraySort.cpp (hunk 16 above)
Comment (duanmeng) [reply velox/functions/prestosql/ArraySort.cpp]: Fixed

File: velox/functions/prestosql/registration/RegistrationFunctions.cpp [hunk 18]
```diff
@@ -108,6 +109,10 @@ void registerMapAllowingDuplicates(
     const std::string& prefix) {
   functions::registerMapAllowingDuplicates(name, prefix);
 }
+
+void registerInternalFunctions(const std::string& prefix) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/registration/RegistrationFunctions.cpp:113]: Remove prefix since it is not used.

File: velox/exec/tests/AggregationFuzzerTest.cpp [hunk 19]
```diff
@@ -40,6 +40,7 @@ DEFINE_string(
 int main(int argc, char** argv) {
   facebook::velox::aggregate::prestosql::registerAllAggregateFunctions();
   facebook::velox::functions::prestosql::registerAllScalarFunctions();
+  facebook::velox::functions::prestosql::registerInternalFunctions();
```
Comment (mbasmanova) [inline velox/exec/tests/AggregationFuzzerTest.cpp:43]: Why can't we register 'canonicalize' function directly?

```
VELOX_REGISTER_VECTOR_FUNCTION(
      udf_$internal$canonicalize, "$internal$canonicalize");
```

**Review by mbasmanova — COMMENTED** [review]
Looks good % a few remaining comments.

File: velox/exec/tests/AggregationFuzzerTest.cpp (hunk 19 above)
Comment (duanmeng) [reply velox/exec/tests/AggregationFuzzerTest.cpp:43]: I came up with the linker error when I used this macro directly here, even though I added the `extern` and `velox_functions_prestosql_impl`.
```Git
+++ b/velox/exec/tests/AggregationFuzzerTest.cpp
+#include "velox/expression/VectorFunction.h"
+extern void registerVectorFunction_udf_$internal$canonicalize(const std::string& name);
 int main(int argc, char** argv) {
   facebook::velox::aggregate::prestosql::registerAllAggregateFunctions();
   facebook::velox::functions::prestosql::registerAllScalarFunctions();
-  facebook::velox::functions::prestosql::registerInternalFunctions();
+  //facebook::velox::functions::prestosql::registerInternalFunctions();
+  VELOX_REGISTER_VECTOR_FUNCTION(
+      udf_$internal$canonicalize, "$internal$canonicalize");

 +++ b/velox/exec/tests/CMakeLists.txt
@@ -197,6 +197,7 @@ add_executable(velox_aggregation_fuzzer_test AggregationFuzzerTest.cpp)

 target_link_libraries(
   velox_aggregation_fuzzer_test
+        velox_functions_prestosql_impl
```

Comment (mbasmanova) [reply velox/exec/tests/AggregationFuzzerTest.cpp:43]: Got it. Then let's leave it as is. Thanks.

File: velox/exec/tests/AggregationFuzzerTest.cpp [hunk 20]
```diff
@@ -80,19 +81,20 @@ int main(int argc, char** argv) {
...
-          {"array_agg", "array_sort({})"},
-          {"array_agg_partial", "array_sort({})"},
-          {"array_agg_merge", "array_sort({})"},
-          {"array_agg_merge_extract", "array_sort({})"},
-          {"set_agg", "array_sort({})"},
-          {"set_union", "array_sort({})"},
-          {"map_agg", "array_sort(map_keys({}))"},
-          {"map_union", "array_sort(map_keys({}))"},
-          {"map_union_sum", "array_sort(map_keys({}))"},
+          {"array_agg", "\"$internal$canonicalize\"({})"},
```
Comment (mbasmanova) [inline velox/exec/tests/AggregationFuzzerTest.cpp:84]: Note that this change reduces coverage in the AggregationFuzzer. Before this change, the Fuzzer was able to verify results against DuckDB (since it supports array_sort function), but now it cannot. We are going to face same problem with using Presto as the reference query runner. A solution could be to change the Fuzzer to apply post-aggregation projections using Velox only (once to Velox results, once to results from reference DB), then compare. This is a follow-up though.
